### ISSUE TITLE
Cloc read file text format UTF8

### ISSUE THREAD

**Issue by Sori31** [AlDanial/cloc#502]
I create file Text with format System.Text.Encoding.UTF8 
   Dim writer As New System.IO.StreamWriter(FileCounter20200715140017.txt, OpenMode.Append, System.Text.Encoding.UTF8)
   For Each value As String In lstFolderFile.Items
        writer.WriteLine(value)
   Next

When using cloc 
   C:\cloc-1.87 --list-file= FileCounter20200715140017.txt,
Message error
![image](https://user-images.githubusercontent.com/62527864/87513913-a6e1b480-c6a3-11ea-98b5-1a3a6ebce89b.png)
[新しいフォルダー.zip](https://github.com/AlDanial/cloc/files/4923419/default.zip)
[FileCounter20200715140017.txt](https://github.com/AlDanial/cloc/files/4923427/FileCounter20200715140017.txt)

The same with folder is
[Điếm LOC.zip](https://github.com/AlDanial/cloc/files/4923450/Di.m.LOC.zip)
[FileCounter20200715140433.txt](https://github.com/AlDanial/cloc/files/4923453/FileCounter20200715140433.txt)

**Comment by AlDanial** [comment]
Unfortunately it will be a while before I can research this.  I'm working long hours at my day job.  Which of your three open issues is most important to you?

**Comment by Sori31** [comment]
I understand, when you have time, please watch it for me.
You preview helped issue #501, thanks you

**Comment by AlDanial** [comment]
I traced the problem to the location where the list file is read.  The string it returns for the SQL file does not match the name of the SQL file in the 'Điếm LOC' folder.  The file's encoding isn't understood for some reason.  I then got a list of all encodings known by Perl on my computer and tried every one with the program below.  None of the encodings worked.  Do you know the Unicode name of the encoding your computer uses?
<pre>
#!/usr/bin/env perl
sub read_list_file {                         # {{{1
    my ($file, $encoding ) = @_;
    # reads filenames from a STDIN pipe if $file == "-"

    print "-> read_list_file($file)\n" if $opt_v > 2;
    my $IN;
    if ($file eq "-") {
        open($IN, $file);
    } else {
        $IN = IO::File->new($file, "<:encoding($encoding)");
    }
    if (!defined $IN) {
        warn "Unable to read $file; ignoring.\n";
        return ();
    }
    my @entry = ();
    while (<$IN>) {
        next if /^\s*$/ or /^\s*#/; # skip empty or commented lines
        s/\cM$//;  # DOS to Unix
        chomp;
        push @entry, $_;
    }
    $IN->close;

    print "<- read_list_file\n" if $opt_v > 2;
    return @entry;
}
# 1}}}

foreach my $enc (qw(
    AdobeStandardEncoding AdobeSymbol AdobeZdingbat
    ascii ascii-ctrl big5-eten big5-hkscs cp1006 cp1026
    cp1047 cp1250 cp1251 cp1252 cp1253 cp1254 cp1255 cp1256
    cp1257 cp1258 cp37 cp424 cp437 cp500 cp737 cp775 cp850
    cp852 cp855 cp856 cp857 cp858 cp860 cp861 cp862 cp863
    cp864 cp865 cp866 cp869 cp874 cp875 cp932 cp936 cp949
    cp950 dingbats euc-cn euc-jp euc-kr gb12345-raw gb2312-raw
    hp-roman8 hz iso-2022-kr iso-8859-1 iso-8859-10 iso-8859-11
    iso-8859-13 iso-8859-14 iso-8859-15 iso-8859-16 iso-8859-2
    iso-8859-3 iso-8859-4 iso-8859-5 iso-8859-6 iso-8859-7
    iso-8859-8 iso-8859-9 iso-ir-165 jis0201-raw jis0208-raw
    jis0212-raw johab koi8-f koi8-r koi8-u ksc5601-raw
    MacArabic MacCentralEurRoman MacChineseSimp MacChineseTrad
    MacCroatian MacCyrillic MacDingbats MacFarsi MacGreek
    MacHebrew MacIcelandic MacJapanese MacKorean MacRoman
    MacRomanian MacRumanian MacSami MacSymbol MacThai MacTurkish
    MacUkrainian MIME-B MIME-Header MIME-Header-ISO_2022_JP
    MIME-Q nextstep null posix-bc shiftjis symbol UCS-2BE
    UCS-2LE UTF-16 UTF-16BE UTF-16LE UTF-32 UTF-32BE UTF-32LE
    UTF-7 utf-8-strict utf8 viscii
)) {
    print "with $enc ======================== \n";
    my @entries = read_list_file("FileCounter20200715140433.txt", $enc);
    my $file_or_dir = @entries[0];
    next unless -r $file_or_dir;
    print "success with $enc";
    last;
}
</pre>

**Comment by Sori31** [comment]
the Unicode name of the encoding your computer : Encoding Japanese "shift-jis", encoding Vietnamese

**Comment by AlDanial** [comment]
I tried these encodings that come with Perl: ``viscii``, ``cp1258``, ``shiftjis`` and also installed ``x-viet-vn`` from 
https://metacpan.org/pod/Encode::VN

None of them worked with your file.  It does not look like a solution exists for your inputs.

**Comment by AlDanial** [comment]
@chris0e3: excellent detective work!  I first encountered the BOM problem in cloc years ago and have a routine, ``read_file()``, that can handle such files.  The failure in this issue was using a simpler routine, ``read_list_file()``, to read the list file.  All I had to do was update ``read_list_file()`` to first call ``read_file()`` then add the extra logic handling STDIN streams and skipping commented lines.
